README.md
# ibanizator [![Tests](https://github.com/softwareinmotion/ibanizator/actions/workflows/ruby.yml/badge.svg)](https://github.com/softwareinmotion/ibanizator/actions/workflows/ruby.yml)

ibanizator calculates the iban for german accounts. The database that is used to convert a bank number to a
BIC is taken from [Deutsche Bundesbank](https://www.bundesbank.de/en/tasks/payment-systems/services/bank-sort-codes/download-bank-sort-codes-626218).

## Installation

Add this line to your application's Gemfile:

```ruby
gem 'ibanizator', git: 'https://github.com/CashCapeAdmin/ibanizator'
```

## Usage

### Calculate IBAN

To calculate the iban for some german accounts, just use this method:

```ruby
ibanizator = Ibanizator.new
ibanizator.calculate_iban country_code: :de, bank_code: '12345678', account_number: '123456789'
```

In the current version the ibanizator gem only works for german accounts.

### Validate an IBAN

To validate the iban you need to check the length and after this check the checksum. For details please refer to
the documentation online (e.g. [Wikipedia](http://es.wikipedia.org/wiki/IBAN)).

This gem provides a simple validator for several contries. All countries that are listed in the Ibanizator::Iban::LENGTHS hash are supported at the moment.

```ruby
iban = Ibanizator.iban_from_string("DE68 2105 0170 0012 3456 78")
iban.valid? # => true
```

### Information provided by an IBAN

The Ibanizator::Iban class provides some handy utility methods to query
information about an iban.

```ruby
iban = Ibanizator.iban_from_string("DE68 2105 0170 0012 3456 78")
iban.country_code # => :DE

# there is extended data for german ibans
iban.extended_data.bank_code        # => "21050170"
iban.extended_data.account_number   # => "12345678"
iban.extended_data.bic              # => "NOLADE21KIE"
```

### Find bank infos

If you need to get a bank name, bank code or a bic from a german bank and you
have either a BIC or a bank code there is a bank db.

```ruby
# find a bank by BIC
bank_1 = Ibanizator.bank_db.bank_by_bic('MARKDEF1100')
bank_1.name       # => 'BBk Berlin'
bank_1.bic        # => 'MARKDEF1100'
bank_1.bank_code  # => '10000000'

# find a bank by bank_code (de: Bankleitzahl)
bank_2 = Ibanizator.bank_db.bank_by_bank_code('100 000 00')
bank_2.name       # => 'BBk Berlin'
bank_2.bic        # => 'MARKDEF1100'
bank_2.bank_code  # => '10000000'

bank_3 = Ibanizator.bank_db.bank_by_bank_code('10000000')
bank_3.name       # => 'BBk Berlin'
bank_3.bic        # => 'MARKDEF1100'
bank_3.bank_code  # => '10000000'

bank_1 == bank_2  # => true
bank_2 == bank_3  # => true

bank_4 = Ibanizator.bank_db.bank_by_bic('OASPDE6AXXX')
bank_4 == bank_2  # => false
```

## Update Banking List from Deutsche Bundesbank

When the Deutsche Bundesbank updates their list, simple download the new `.txt` file and replace the file located at `https://github.com/CashCapeAdmin/ibanizator/blob/master/db/blz.txt` from here [Bundesbank](https://www.bundesbank.de/en/tasks/payment-systems/services/bank-sort-codes/download-bank-sort-codes-626218).

Create a new minor release and update the gem in the project. 

## Licence

The code is availiable under the MIT-Licence


CHANGELOG.md
# Changelog

## 0.4.15

- **Features:**
  - Updated bank database (valid until March 04 2023)

## 0.4.14

- **Features:**
  - Updated bank database (valid until December 04 2022)

## 0.4.13

- **Features:**
  - Added 3.1 to tested ruby versions
  - Updated bank database (valid until September 04 2022)

## 0.4.12

- **Features:**
  - Updated bank database (valid until December 05 2021)

## 0.4.11

- **Features:**
  - Updated bank database (valid until September 07 2021)

## 0.4.10

- **Features:**
  - Updated blz.txt (valid until 06.06.2021)
  - Added 3.0.0 to tested ruby versions
  - Added jruby, truffleruby and truffleruby-head to tested ruby versions
  - Changed gem to be tested via github actions instead of travis.ci

## 0.4.9

- **Features:**
  - Updated blz.txt (valid until 07.03.2021)
  - Update tested ruby versions to 2.7.2

## 0.4.8

- **Features:**
  - Updated blz.txt (valid until 06.12.2020)

## 0.4.7

- **Features:**
  - Updated blz.txt (valid until 06.09.2020)
  - Add ruby version 2.7.1 to tests
  - Update tested ruby versions to 2.5.8 and 2.6.6
  - Removed support for ruby 2.4 due to eol since 31.03.2020
  - Updated dev dependencies
  - Updated and added multiple iban lengths ([#27](https://github.com/softwareinmotion/ibanizator/pull/27), [#28](https://github.com/softwareinmotion/ibanizator/pull/28)) ~ *thanks to [skorth](https://github.com/skorth) and [svoop](https://github.com/svoop)*

## 0.4.6

- **Features:**
  - Updated blz.txt (valid until 08.03.2020)
  - Update tested ruby versions to 2.4.9, 2.5.7 and 2.6.5

## 0.4.5

- **Features:**
  - Updated blz.txt (valid until 08.12.2019)
  - Update tested ruby versions to 2.4.7, 2.5.6 and 2.6.4

## 0.4.4

- **Features:**
  - Updated blz.txt (valid until 08.09.2019)
  - Update tested ruby versions to 2.4.6, 2.5.5 and 2.6.3
  - Removed support for ruby 2.3 due to eol since 31.03.2019
  - Removed support for ruby 2.2 due to eol since 31.03.2018

## 0.4.3

- **Features:**
  - Updated blz.txt (valid until 02.06.2019)
  - Updated ruby version from 2.5.3 to 2.6.0

## 0.4.2

- **Features:**
  - Updated blz.txt (valid until 03.03.2019)
  - Updated ruby version from 2.5.1 to 2.5.3
  - Updated Bundesbank link

## 0.4.1

- **Features:**
  - Updated blz.txt (valid until 02.12.2018)

## 0.4.0

- **Breaking Changes:**
  - Removed deprecated class `SwiftBic::BankDb` in favor of `Ibanizator::BankDb` ([#24](https://github.com/softwareinmotion/ibanizator/pull/24))
  - Removed deprecated method `Iban::Validator.validate_iban` in favor of `Ibanizator.iban_from_string(iban).valid?` ([#24](https://github.com/softwareinmotion/ibanizator/pull/24))
  - Removed deprecated method `Ibanizator.new.validate_iban(iban)` in favor of `Ibanizator.iban_from_string(iban).valid?` ([#24](https://github.com/softwareinmotion/ibanizator/pull/24))
  - Removed deprecated method `Ibanizator.new.bic(bank_code)` in favor of `Ibanizator::BankDb.new.bank_by_bank_code(bank_code).bic` ([#24](https://github.com/softwareinmotion/ibanizator/pull/24))
  - Removed deprecated method `Ibanizator.new.bank_name(bank_code)` in favor of `Ibanizator::BankDb.new.bank_by_bank_code(bank_code).name` ([#24](https://github.com/softwareinmotion/ibanizator/pull/24))

- **Features:**
  - Added `formatted_iban_string` instance method to `Ibanizator::Iban` for an iban with spaces. ([#17](https://github.com/softwareinmotion/ibanizator/pull/17)) ~ *thanks to [phoet](https://github.com/phoet)*
  - Also adding a `bank_name` and `to_hash` instance method to `Ibanizator::Iban::ExtendedData::DE`. ([#17](https://github.com/softwareinmotion/ibanizator/pull/17)) ~ *thanks to [phoet](https://github.com/phoet)*
  - Added short version of bics containing an optional branch code e.g. Deutsche Bank can be found either with DEUTDEBBXXX or DEUTDEBB. Additionally changed `#bank_by_bank_code` to favor banks whose bic ends in 'XXX' ([#22](https://github.com/softwareinmotion/ibanizator/pull/22))
  - Removed support for ruby 2.1 and added support for ruby 2.4 and 2.5 ([#23](https://github.com/softwareinmotion/ibanizator/pull/23))
  - Changed the contact information ([#25](https://github.com/softwareinmotion/ibanizator/pull/25))

## 0.3.5

- **Features:**
  - Updated blz.txt (valid until 02.09.2018)

- **Fixes:**
  - Fixed failing test suite ([#20](https://github.com/softwareinmotion/ibanizator/pull/20))

## 0.3.4 and blow

~ *Here be dragons* ~


lib/ibanizator/iban/lengths.rb
class Ibanizator
  class Iban
    # according to wikipedia (http://en.wikipedia.org/wiki/IBAN)
    LENGTHS = {
      :AL => 28,
      :AD => 24,
      :AT => 20,
      :AZ => 28,
      :BH => 22,
      :BY => 28,
      :BE => 16,
      :BA => 20,
      :BR => 29,
      :BG => 22,
      :CR => 22,
      :HR => 21,
      :CY => 28,
      :CZ => 24,
      :DK => 18,
      :DO => 28,
      :TL => 23,
      :EG => 27,
      :EE => 20,
      :FO => 18,
      :FI => 18,
      :FR => 27,
      :GE => 22,
      :DE => 22,
      :GI => 23,
      :GR => 27,
      :GL => 18,
      :GT => 28,
      :HU => 28,
      :IS => 26,
      :IQ => 23,
      :IE => 22,
      :IL => 23,
      :IT => 27,
      :JO => 30,
      :KZ => 20,
      :XK => 20,
      :KW => 30,
      :LV => 21,
      :LB => 28,
      :LI => 21,
      :LT => 20,
      :LU => 20,
      :MK => 19,
      :MT => 31,
      :MR => 27,
      :MU => 30,
      :MC => 27,
      :MD => 24,
      :ME => 22,
      :NL => 18,
      :NO => 15,
      :PK => 24,
      :PS => 29,
      :PL => 28,
      :PT => 25,
      :QA => 29,
      :RO => 24,
      :LC => 32,
      :SM => 27,
      :SA => 24,
      :RS => 22,
      :SC => 31,
      :SK => 24,
      :SI => 19,
      :ES => 24,
      :SE => 24,
      :CH => 21,
      :TN => 24,
      :TR => 26,
      :UA => 29,
      :AE => 23,
      :GB => 22,
      :VA => 22,
      :VG => 24
    }.freeze
  end
end
